File: mastra_agent/src/metrics.ts
import { MeterProvider } from '@opentelemetry/sdk-metrics';
import { PrometheusExporter } from '@opentelemetry/exporter-prometheus';
import { resourceFromAttributes } from '@opentelemetry/resources';
import { ATTR_SERVICE_NAME } from '@opentelemetry/semantic-conventions';

/**
 * OpenTelemetry metrics for the Mastra agent (this service, mastra_agent).
 *
 * Prefixed `mastra_`, matching demo_mcp_gateway's `mcpgw_` convention
 * (prom-client there, OTel Metrics API here — both still end up scraped by
 * Prometheus the same way, see monitoring/prometheus.yml).
 *
 * Deliberately does NOT touch handleRun's control flow, error handling, or
 * return values — index.ts wraps the existing function with pure observation
 * (record on settle, re-throw unchanged). Note: runHandler's own try/catch
 * swallows agent/tool/LLM errors into an AG-UI RUN_ERROR SSE event rather
 * than rejecting, so mastra_run_errors_total only counts something escaping
 * that catch entirely (Express/middleware-level failures) — not agent-level
 * failures, which the run always reports as a "successful" HTTP response.
 */

// preventServerStart: this service already listens on cfg.port (see index.ts) —
// metrics are served from that SAME port via exporter.getMetricsRequestHandler,
// not a second exporter-owned server.
const exporter = new PrometheusExporter({ preventServerStart: true });
const resource = resourceFromAttributes({
  [ATTR_SERVICE_NAME]: process.env.OTEL_SERVICE_NAME || 'mastra-agent',
});
const meterProvider = new MeterProvider({ readers: [exporter], resource });
const meter = meterProvider.getMeter('mastra-agent');

export const runDuration = meter.createHistogram('mastra_run_duration_seconds', {
  description: 'Duration of /run agent invocations',
  unit: 's',
});

export const runErrors = meter.createCounter('mastra_run_errors_total', {
  description: 'Agent runs that threw before runHandler\'s own error handling could respond',
});

export { exporter };

File: openai_agent/src/metrics.py
"""OpenTelemetry metrics for the OpenAI agent (this service, openai_agent).

Prefixed `openaiagent_`, matching demo_mcp_gateway's `mcpgw_` convention
(prom-client there, OTel Metrics API here — both still end up scraped by
Prometheus the same way, see monitoring/prometheus.yml).

Deliberately does NOT touch _stream's control flow, error handling, or
yielded events — run_handler.py wraps the existing generator with pure
observation (record on settle, re-raise unchanged). Note: _stream's own
try/except swallows agent/tool/LLM errors into an AG-UI RUN_ERROR SSE
event rather than raising, so openaiagent_run_errors_total only counts
something escaping that except entirely — not agent-level failures, which
the run always reports as a "successful" HTTP response.
"""

import os

from opentelemetry.exporter.prometheus import PrometheusMetricReader
from opentelemetry.sdk.metrics import MeterProvider
from opentelemetry.sdk.resources import Resource

# PrometheusMetricReader self-registers into prometheus_client's default
# global REGISTRY, so main.py's /metrics route can serve it with a plain
# prometheus_client.generate_latest() call — no second exporter-owned server.
_resource = Resource.create({"service.name": os.environ.get("OTEL_SERVICE_NAME", "openai-agent")})
_reader = PrometheusMetricReader()
_provider = MeterProvider(metric_readers=[_reader], resource=_resource)
_meter = _provider.get_meter("openai-agent")

run_duration = _meter.create_histogram(
    "openaiagent_run_duration_seconds",
    unit="s",
    description="Duration of /run agent invocations",
)

run_errors = _meter.create_counter(
    "openaiagent_run_errors_total",
    description="Agent runs that raised before _stream's own error handling could respond",
)
